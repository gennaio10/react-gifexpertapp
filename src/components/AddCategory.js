import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleOnSubmit = (e) => {
    //evita que se recarga la pagina
    e.preventDefault();

    if (inputValue.trim().length > 1) {
      setCategories((categoriesE) => [inputValue, ...categoriesE]);
      setInputValue("");
    }
  };
  
  return (
    <form onSubmit={handleOnSubmit}>
      <h2>Add Category</h2>
      <input id="icategoria" type="text" value={inputValue} onChange={handleChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
