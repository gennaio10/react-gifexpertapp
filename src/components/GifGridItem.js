import React from "react";
import PropTypes from 'prop-types';

export const GifGridItem = ({ title, url }) => {
  return (
    <div id="dimg" className="card animate__animated animate__bounce">
      <img id="isrc" src={url} alt={title} />
      <p id="ptitle">{title}</p>
    </div>
  );
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};



