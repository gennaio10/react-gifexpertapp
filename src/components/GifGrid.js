import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from "prop-types";

export const GifGrid = ({ category }) => {
  const { data:imagenes, loadind } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {loadind && <p>Loading...</p>}
      <div className="card-grid">
        {imagenes.map((imagenConUrl) => (
          <GifGridItem key={imagenConUrl.id} {...imagenConUrl} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
