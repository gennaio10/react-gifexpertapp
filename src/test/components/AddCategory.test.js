import React from "react";
import "@testing-library/jest-dom";
import { AddCategory } from "../../components/AddCategory";
import { shallow } from "enzyme";

describe("Pruebas de <AddCategory />", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);
  const value = "Texto de prueba";

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de cambiar el texto de la caja de texto", () => {
    const icategoria = wrapper.find("#icategoria");
    icategoria.simulate("change", { target: { value } });
    expect(wrapper).toMatchSnapshot();
  });

  test("No debe postear la informacion onSubmit", () => {
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    // Se espera no se llame la funcion "setCategories"
    expect(setCategories).not.toHaveBeenCalled();
  });

  test("debe llamar a setCategories y limpiar la caja de texto", () => {
    const icategoria = wrapper.find("#icategoria");
    icategoria.simulate("change", { target: { value } });
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    // Se espera se llame la funcion "setCategories"
    expect(setCategories).toHaveBeenCalled();

    // Se espera se halla llamado "setCategories" solo una vez
    expect(setCategories).toHaveBeenCalledTimes(1);

    // Se espera se llame la funcion "setCategories" con una funcion como parametro
    // ASI ESTA OK    setCategories((categoriesE) => [inputValue, ...categoriesE]);
    // ASI ESTA MAL   setCategories(12323233);
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

    // el valor del input debe estar limpio
    expect(wrapper.find("#icategoria").props().value).toBe('');
  });
});
