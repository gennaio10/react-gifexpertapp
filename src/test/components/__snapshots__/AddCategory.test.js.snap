// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <AddCategory /> debe de cambiar el texto de la caja de texto 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Add Category
  </h2>
  <input
    id="icategoria"
    onChange={[Function]}
    type="text"
    value="Texto de prueba"
  />
</form>
`;

exports[`Pruebas de <AddCategory /> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Add Category
  </h2>
  <input
    id="icategoria"
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
